File: 0130-Surrounded-Regions.py
class Solution:
    def solve(self, board: List[List[str]]) -> None:
        ROWS, COLS = len(board), len(board[0])
        
        def capture(r, c):
            if (r < 0 or c < 0 or r == ROWS or c == COLS 
                or board[r][c] != "O"):
                return
            board[r][c] = "T"
            capture(r + 1, c)
            capture(r - 1, c)
            capture(r, c + 1)
            capture(r, c - 1)
        
        # 1. (DFS) Capture unsurrounded regions (O -> T)
        for r in range(ROWS):
            for c in range(COLS):
                if (board[r][c] == "O" and
                    (r in [0, ROWS - 1] or c in [0, COLS - 1])):
                    capture(r, c)
        
        # 2. Capture surrounded regions (O -> X)
        for r in range(ROWS):
            for c in range(COLS):
                if board[r][c] == "O":
                    board[r][c] = "X"
                    
        # 3. Uncapture unsurrounded regions (T -> O)
        for r in range(ROWS):
            for c in range(COLS):
                if board[r][c] == "T":
                    board[r][c] = "O"

File: swift/0130-surrounded-regions.swift
class Solution {
    func solve(_ board: inout [[Character]]) {
        var rows = board.count
        var columns = board[0].count
        
        // 1. DFS Capture unsurrounded regions (O -> T)
        for r in 0..<rows {
            for c in 0..<columns {
                if board[r][c] == "O" &&
                    ([0, rows - 1].contains(r) || [0, columns - 1].contains(c)) {
                    capture(r, c, &board)
                }
            }
        }
        
        // 2. Capture surrounded regions (O -> X)
        for r in 0..<rows {
            for c in 0..<columns {
                if board[r][c] == "O" {
                    board[r][c] = Character("X")
                }
            }
        }
        
        // 3. Uncapture surrounded regions (T -> O)
        for r in 0..<rows {
            for c in 0..<columns {
                if board[r][c] == "T" {
                    board[r][c] = Character("O")
                }
            }
        }
    }
    
    func capture(_ r: Int, _ c: Int, _ board: inout [[Character]]) {
        // Make sure r and c are in bounds
        guard r >= 0, r < board.count, c >= 0, c < board[0].count, board[r][c] == "O" else {
            return
        }
        
        board[r][c] = Character("T")

        capture(r - 1, c, &board)
        capture(r + 1, c, &board)
        capture(r, c - 1, &board)
        capture(r, c + 1, &board)
    }
}